### ISSUE TITLE
win_acl module cannot modify ACLs for registry keys

### ISSUE THREAD

**Issue by dagwieers** [ansible/ansible#25474]
_From @bpleines on October 19, 2016 2:26_

##### ISSUE TYPE

- Feature Idea
##### COMPONENT NAME

win_acl module
##### ANSIBLE VERSION

```
ansible 2.3.0 (devel 05aed6e52e) last updated 2016/10/18 19:10:43 (GMT -700)
  lib/ansible/modules/core: (detached HEAD b59b5d36e0) last updated 2016/10/18 19:10:47 (GMT -700)
  lib/ansible/modules/extras: (detached HEAD 3f77bb6857) last updated 2016/10/18 19:10:51 (GMT -700)
  config file = 
  configured module search path = Default w/o overrides
```
##### OS / ENVIRONMENT

Control Node: Fedora 23
Host Machine: AWS Windows 2012 Member Server
##### SUMMARY

The win_acl module currently does not support modification of ACLs for registry keys - only for folders and files. There is not currently a module to modify ACLs for registry keys, which are used for customization/hardening of Windows OSes. This is needed for an ongoing Windows STIG Remediation engagement.
##### STEPS TO REPRODUCE

run the following tasks below

```
- name: "V-39331 dummy"
  win_acl:
     path: '\Windows\SYSVOL'
     user: 'BUILTIN\Administrators'
     rights: 'FullControl'
     type: 'allow'
     state: 'present'
     inherit: 'ContainerInherit, ObjectInherit'
     propogation: 'InheritOnly'

- name: "V-1152 dummy"
  win_acl:
     path: 'HKLM:\SYSTEM\CurrentControlSet\Control\SecurePipeServers\Winreg\'
     user: 'BUILTIN\Administrators'
     rights: 'FullControl'
     type: 'allow'
     state: 'present'
     inherit: 'ContainerInherit, ObjectInherit'
     propagation: 'None'
```

##### EXPECTED RESULTS

The task list runs cleanly, successfully updating each ACE.
[ansiblePatchedModules.zip](https://github.com/ansible/ansible-modules-extras/files/537808/ansiblePatchedModules.zip)
[ansiblePatchedModules.zip](https://github.com/ansible/ansible-modules-extras/files/537809/ansiblePatchedModules.zip)
##### ACTUAL RESULTS

The first task runs fine because it is a folder, but the second task fails because it is a registry key. This is because under the hood permissions for registry keys are set with "New-Object System.Security.AccessControl.RegistryAccessRule" while folder and files are set with "New-Object System.Security.AccessControl.FileSystemAccessRule." Incorporating this functionality into the win_acl.ps1 and win_acl_inheritance.ps1 would involve patching both modules. I have attached _very_ basic patches so that the task list provided above works for both tasks.

```
fatal: [54.71.248.237]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_name": "win_acl"
    }, 
    "msg": "an error occured when attempting to present FullControl permission(s) on HKLM:\\SYSTEM\\CurrentControlSet\\Control\\SecurePipeServers\\Winreg\\ for BUILTIN\\Administrators"
}
```

_Copied from original issue: ansible/ansible-modules-extras#3233_

**Comment by dagwieers** [comment]
_From @ansibot on October 19, 2016 2:32_



@schwartzmx, @trondhindenes, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dagwieers** [comment]
_From @bpleines on October 19, 2016 2:40_

I listed this as a feature idea and not a bug because the module does not state that it has registry key functionality

**Comment by dagwieers** [comment]
_From @ansibot on December 7, 2016 14:34_

This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by ansibot** [comment]
cc @h0nIg
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jborean93** [comment]
While it does not say so, win_acl does have the ability to modify the ACL for registry keys. I will create a PR to update the documentation but I have first tested the following task with success.

```
  - win_acl:
     path: 'HKLM:\SYSTEM\Test'
     user: 'BUILTIN\Users'
     rights: 'FullControl'
     type: 'allow'
     state: 'present'
     inherit: 'ContainerInherit, ObjectInherit'
     propagation: 'None'
```

**Comment by bpleines** [comment]
@jborean, it has this capability because the patch that I submitted above fixed the 'issue'. At the time your above task wouldn't work, but the docs did need to be updated (thank you!)

**Comment by jborean93** [comment]
Thanks @bpleines I was unsure who raised the original PR to add that functionality but thank you for doing it.

**Comment by jborean93** [comment]
Closing issue as changes have been merged into devel. Thanks @bpleines for raising the issue and putting in the functionality.
